GDP-2: Regions (#24)
GridTools/gt4py: docs/gt4py/GDPs/gdp-0002-regions.rst

        if grid.west_edge:​
            x_edge_ub(ut, ub, dt5=dt5, ​origin=(grid.local_istart, grid.local_jstart, 0), ​domain=domain_x_edge)​
        if grid.south_edge:
            y_edge(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jstart, 0), domain=domain_y_edge)
        if grid.north_edge:
            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
        if grid.east_edge:
            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)

The specific feature that we are proposing is the addition of a ``with horizontal()`` context that specifies the horizontal iteration space (over parallel axes) using ``region`` objects.
The arguments of ``region`` object's ``__getitem__`` method use *axis offsets* to define a subregion of the stencil computational domain in the parallel ``I`` and ``J`` axes.

Using this specification, the example is transformed into (for an ``NxN`` processor layout):

.. code-block:: python

    istart = I[0] - np_local * (procid % N)
    jstart = J[0] - np_local * (procid / N)
    iend = I[-1] + np_global - (procid % N) * np

>     jend = J[-1] + np_global - (procid / N) * np

Reviewer: Shouldn't `np` here be `np_local` or am I missing something?

    @gtscript.stencil()
    def stencil(uc: Field, vc: Field, cosa: Field, rsina: Field, ut: Field, ub: Field, dt5: float, dt4: float):
        from __externals__ import istart, iend, jstart, jend
        with computation(PARALLEL), interval(...):
            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
            with horizontal(region[istart, :], region[iend, :]):
                ub = dt5 * (ut[0, -1, 0] + ut)
            with horizontal(region[:, jstart], region[:, jend]):
                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])

> where ``np_local`` and ``np_global`` are the local and global number of horizontal cells in either direction.

Reviewer: I spent more minutes than I would like to confess trying to guess the meaning of `np_global` and `np_local`... :facepalm:  I would suggest to move the definition of both terms before the actual source code example.

    with computation(PARALLEL), interval(...):
        with horizontal(region[:I[0], :]):
            u = 0

This will set ``u=0`` in all extended computation points to the left of the compute domain.

Examples of this are shown in the image below.
The blue line shows the compute domain along the ``I`` axis, and two examples of region axis slices are shown in red.

.. image:: _static/gdp-0002/axis_offsets.svg

Execution
+++++++++

Another key feature to remember when using regions is that these should be thought of as specifying specialized computation at points.
These are therefore not guaranteed to execute, except where inside the compute domain.
The statements inside a block with ``region[:I[0]-1, :]`` will only execute where the outputs from that block are necessary to compute something else with an extent.
For example, the following will execute

.. code:: python

    with computation(PARALLEL), interval(...):
        with horizontal(region[I[0]-1, :]):
            field_in = 0.0
        field_out = field_in[-1, 0, 0] + field_in[0, 0, 0]

since the ``field_in`` value at ``I[0]-1`` is being consumed to compute a value of an output field inside the compute domain.

> If this used ``I[0]-2``, the code would be ignored.

Reviewer: ```suggestion
If the region were defined using ``I[0]-2``, the code would be ignored.
```
